ikarus/finiteelements/mechanics/kirchhoffloveshell.hh

```diff
@@ -180,30 +203,179 @@ namespace Ikarus {
+      const Eigen::Matrix<ScalarType, 3, 2> gradu = toEigen(uFunction.evaluateDerivative(
+          gpIndex, Dune::wrt(spatialAll), Dune::on(Dune::DerivativeDirections::referenceElement)));
+      const Eigen::Matrix<ScalarType, 2, 3> j     = J + gradu.transpose();
+      const Eigen::Vector3<ScalarType> a3N        = (j.row(0).cross(j.row(1)));
+      const Eigen::Vector3<ScalarType> a3         = a3N.normalized();
+      Eigen::Vector<ScalarType, 3> bV             = h * a3;
+      bV(2) *= 2;  // Voigt notation requires the two here
+      const auto BV     = toVoigt(toEigen(geo.impl().secondFundamentalForm(gpPos)));
+      const auto kappaV = (BV - bV).eval();
+      return std::make_tuple(C, epsV, kappaV, j, J, h, H, a3N, a3);
+    }
+
+    template <typename ScalarType>
+    void calculateMatrixImpl(const FERequirementType &par, typename Traits::template MatrixType<ScalarType> K,
+                             const std::optional<const Eigen::VectorX<ScalarType>> &dx = std::nullopt) const {
+      K.setZero();
+      using namespace Dune::DerivativeDirections;
+      using namespace Dune;
+      const auto uFunction = getDisplacementFunction(par, dx);
+      const auto &lambda   = par.getParameter(FEParameter::loadfactor);
+      const auto geo       = this->localView().element().geometry();
+      this->membraneStrain.pre(geo, uFunction);
+
+      // Internal forces
+      for (const auto &[gpIndex, gp] : uFunction.viewOverIntegrationPoints()) {
+        const auto intElement = geo.integrationElement(gp.position()) * gp.weight();
+        const auto [C, epsV, kappaV, jE, J, h, H, a3N, a3]
+            = computeMaterialAndStrains(gp.position(), gpIndex, geo, uFunction);
+        const Eigen::Vector<ScalarType, 3> membraneForces = thickness_ * C * epsV;
+        const Eigen::Vector<ScalarType, 3> moments        = Dune::power(thickness_, 3) / 12.0 * C * kappaV;
+
+        const auto &Nd  = localBasis.evaluateJacobian(gpIndex);
+        const auto &Ndd = localBasis.evaluateSecondDerivatives(gpIndex);
+        for (size_t i = 0; i < numberOfNodes; ++i) {
+          Eigen::Matrix<ScalarType, 3, 3> bopIMembrane
+              = membraneStrain.derivative(gp.position(), jE, Nd, geo, uFunction, localBasis, i);
+          Eigen::Matrix<ScalarType, 3, 3> bopIBending = bopBending(jE, h, Nd, Ndd, i, a3N, a3);
+          for (size_t j = i; j < numberOfNodes; ++j) {
+            Eigen::Matrix<ScalarType, 3, 3> bopJMembrane
+                = membraneStrain.derivative(gp.position(), jE, Nd, geo, uFunction, localBasis, j);
+            Eigen::Matrix<ScalarType, 3, 3> bopJBending = bopBending(jE, h, Nd, Ndd, j, a3N, a3);
+            K.template block<3, 3>(3 * i, 3 * j)
+                += thickness_ * bopIMembrane.transpose() * C * bopJMembrane * intElement;
+            K.template block<3, 3>(3 * i, 3 * j)
+                += Dune::power(thickness_, 3) / 12.0 * bopIBending.transpose() * C * bopJBending * intElement;
```

Comment:
`3` here can be replaced with `Traits::worlddim` as done in the `getDisplacementFunction` such that it is not hard coded here?